types/index.d.ts

```diff
@@ -0,0 +1,54 @@
+// TypeScript Version: 3.0
+
+import {Node, Parent} from 'unist'
+import {Test} from 'unist-util-is'
+
+declare namespace mdastUtilToc {
+  interface Link extends Parent {
+    type: 'link'
+    title: null
+    url: string
+    children: Node[]
+  }
+
+  interface Paragraph extends Parent {
+    type: 'paragraph'
+    children: Link[]
+  }
+
+  interface ListItem extends Parent {
+    type: 'listItem'
+    loose: boolean
+    spread: boolean
+    children: Node[]
+  }
+
+  interface List extends Parent {
+    type: 'list'
+    ordered: boolean
+    spread: boolean
+    children: ListItem[]
+  }
+
+  interface TOCOptions {
+    heading?: string
+    maxDepth?: number
+    skip?: string
+    tight?: boolean
+    prefix?: string
+    parents?: Test<Node> | Array<Test<any>>
+  }
```

Comment:
TSDocs for each of the settings, based off
 
###

> ## `options`
> 
> ###### `options.heading`
> 
> [Heading][] to look for (`string`), wrapped in `new RegExp('^(' + value + ')$',
> 'i')`.
> 
> ###### `options.maxDepth`
> 
> Maximum heading depth to include in the table of contents (`number`, default:
> `6`),
> This is inclusive: when set to `3`, level three headings are included (those
> with three hashes, `###`).
> 
> ###### `options.skip`
> 
> Headings to skip (`string`, optional), wrapped in
> `new RegExp('^(' + value + ')$', 'i')`.
> Any heading matching this expression will not be present in the table of
> contents.
> 
> ###### `options.tight`
> 
> Whether to compile list-items tightly (`boolean?`, default: `false`).
> 
> ###### `options.prefix`
> 
> Add a prefix to links to headings in the table of contents (`string?`,
> default: `null`).
> Useful for example when later going from [mdast][] to [hast][] and sanitizing
> with [`hast-util-sanitize`][sanitize].
> 
> ###### `options.parents`
> 
> Allows headings to be children of certain node [type][]s (default: the to `toc`
> given `tree`, to only allow top-level headings).
> Internally, uses [unist-util-is][is] to check, so `parents` can be any
> [`is`-compatible][is] test.
> 
> For example, this would allow headings under either `root` or `blockquote` to be
> used:
> 
> ```js
> toc(tree, {parents: ['root', 'blockquote']})
> ```

Could be a nice addition for the developer experience